### PR TITLE
Mocha test utilities for add-on code (#1082)

### PR DESCRIPTION
**Pull request by k88hudson**
This patch adds some test utilities for testing add-on code in mocha. It includes:

- a `shims/` folder with a bunch of shims for `chrome`, `sdk/tabs`, and others, including tests for those shims (in `content-test/meta`)
- updates for karma to take the shims into account, and an injector loader for overriding require
- re-written tests for `PrefsProvider` using the new strategy/shims
- documentation / a guide

---
This change is [Reviewable](https://reviewable.io/reviews/mozilla/activity-stream/1160)

### CODE REVIEW

**Comment by k88hudson** [comment]
@sarracini could you take a look?

File: shims/sdk/tabs.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+      window: {},
+      readyState: "complete"
+    }, custom);
+    Object.keys(props).forEach(key => {
+      this[key] = props[key];
+    });
+
+    // Methods
+    this.pin = sinon.spy();
+    this.pin = sinon.spy();
```
Comment (sarracini) [inline shims/sdk/tabs.js]: dupe?

File: content-test/meta/shims/EventEmitter.test.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
+const EventEmitter = require("shims/_utils/EventEmitter");
+const noop = () => {};
+
+describe("EventEmitter shim", () => {
```
Comment (sarracini) [inline content-test/meta/shims/EventEmitter.test.js]: nit: just for consistency with other shim tests, can you drop the "shim" in the name?

File: content-test/addon-tests-how-to.md [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+- relative modules are supported (e.g. `require("./foo.js")` will resolve to `foo.js` in the same directory as the file that calls it)
+- absolute paths are resolved relative to the root directory (e.g. `require("foo.js")` will resolve to `activity-stream/foo.js`)
+- some modules can be loaded from Firefox by name, such as `chrome` and `sdk` (e.g. `require("sdk/tabs")`)
+- modules **cannot** be loaded directly from npm. external dependencies need to be added to `addon/vendor-src.js` and imported with `require("addon/vendor.bundle.js")`
+
+### Module loading in content code
+
+All content-side javascript (with the exception of frame scripts) is preprocessed by `webpack`, which is what provides its module loader. It has the following properties:
+
+- relative modules are supported (same as add-ok sdk)
```
Comment (sarracini) [inline content-test/addon-tests-how-to.md]: typo nit: add-on 😶

**Comment by sarracini** [comment]
this is wonderful, just a couple tiny nits and a commit squash and it's an r+
💃 💃 💃

**Comment by pdehaan** [comment]
Not sure that it adds anything of great value, but I like ESLint and just happened to be staring at this module with my special eyes this morning: https://www.npmjs.com/package/eslint-plugin-mocha


```diff
diff --git a/.eslintrc b/.eslintrc
index 41b3720..9a1a940 100644
--- a/.eslintrc
+++ b/.eslintrc
@@ -21,6 +21,7 @@
   },
   "plugins": [
     "json",
+    "mocha",
     "mozilla",
     "react"
   ],
@@ -28,6 +29,17 @@
     "eslint:recommended"
   ],
   "rules": {
+    "mocha/handle-done-callback": 1,
+    "mocha/no-exclusive-tests": 1,
+    "mocha/no-global-tests": 1,
+    "mocha/no-mocha-arrows": 0,
+    "mocha/no-pending-tests": 1,
+    "mocha/no-return-and-callback": 1,
+    "mocha/no-skipped-tests": 1,
+    "mocha/no-top-level-hooks": 0,
+    "mocha/valid-suite-description": 1,
+    "mocha/valid-test-description": 1,
+
     "mozilla/components-imports": 1,
     "mozilla/import-globals-from": 1,
     "mozilla/this-top-level-scope": 1,
diff --git a/dev-prefs.json b/dev-prefs.json
index 91f5d64..47af20f 100644
--- a/dev-prefs.json
+++ b/dev-prefs.json
```

I ran this against **master** (and not this branch), but it only spat out these 4 relatively boring warnings (all of them [**mocha/valid-test-description**](https://github.com/lo1tuma/eslint-plugin-mocha/blob/master/docs/rules/valid-test-description.md) rule violations [because the test didn't specifically start with the word "should"]):

```sh
$ npm run test:lint

> activity-streams@1.1.2 test:lint /Users/pdehaan/dev/github/mozilla/activity-streams
> eslint --ext=.js,.jsx,.json . .storybook && sass-lint -v -q


/Users/pdehaan/dev/github/mozilla/activity-streams/content-test/components/SiteIcon.test.js
  60:5  warning  Invalid "it()" description found  mocha/valid-test-description

/Users/pdehaan/dev/github/mozilla/activity-streams/content-test/faker.test.js
  75:3  warning  Invalid "it()" description found  mocha/valid-test-description

/Users/pdehaan/dev/github/mozilla/activity-streams/content-test/selectors/siteMetadataSelectors.test.js
  63:5  warning  Invalid "it()" description found  mocha/valid-test-description
  67:5  warning  Invalid "it()" description found  mocha/valid-test-description

✖ 4 problems (0 errors, 4 warnings)
```

**Comment by k88hudson** [comment]
@pdehaan seems good to me! do you want to put that up as a PR?

**Pull request merged**